internal/dot/graph.go

```diff
@@ -0,0 +1,319 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dot
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// ErrorType of a constructor or group is updated when they fail to build.
+type ErrorType int
+
+const (
+	noError ErrorType = iota
+	rootCause
+	transitiveFailure
+)
+
+// CtorID is a unique numeric identifier for constructors.
+type CtorID uintptr
+
+// Ctor encodes a constructor provided to the container for the DOT graph.
+type Ctor struct {
+	Name        string
+	Package     string
+	File        string
+	Line        int
+	ID          CtorID
+	Params      []*Param
+	GroupParams []*Group
+	Results     []*Result
+	ErrorType   ErrorType
+}
+
+// Node is a single node in a graph and is embedded into Params and Results.
+type Node struct {
+	Type  reflect.Type
+	Name  string
+	Group string
+}
+
+// Param is a parameter node in the graph.
+type Param struct {
+	*Node
+
+	Optional bool
+}
+
+// Result is a result node in the graph.
+type Result struct {
+	*Node
+
+	// GroupIndex is added to differentiate grouped values from one another.
+	// Since grouped values have the same type and group, their Node / string
+	// representations are the same so we need indices to uniquely identify
+	// the values.
+	GroupIndex int
+}
+
+// Group is a group node in the graph.
+type Group struct {
+	// Type is the type of values in the group.
+	Type      reflect.Type
+	Name      string
+	Results   []*Result
+	ErrorType ErrorType
+}
+
+// Graph is the DOT-format graph in a Container.
+type Graph struct {
+	Ctors   []*Ctor
+	ctorMap map[CtorID]*Ctor
+
+	Groups   []*Group
+	groupMap map[groupKey]*Group
+
+	Failed *FailedNodes
+}
+
+// FailedNodes is the nodes that failed in the graph.
+type FailedNodes struct {
+	// RootCauses is a list of the point of failures. They are the root causes
+	// of failed invokes and can be either missing types (not provided) or
+	// error types (error providing).
+	RootCauses []*Result
```

Comment:
I could be mistaken, but it appears that there will only ever be a single `RootCause` - appending to the internal slice is guarded by a `isRootCause := len(dg.Failed.RootCauses) == 0`. I don't think we need a slice in this case.

This is not at all actionable - things will work exactly the same regardless :)